File: src/services/webrtc/voiceHealth.ts
/** Counters, not audio loudness: silence/DTX and intentional mute are not failures. */
export class VoiceHealth {
  private previous?: { received: number; sent: number | null };
  private suspectSince: number | null = null;
  private remoteProgressAt = -Infinity;
  private lastReason: string | null = null;
  private nextRepairAt = 0;
  private failures = 0;
  private healthySince: number | null = null;

  resetSample() {
    this.previous = undefined;
    this.suspectSince = null;
    this.remoteProgressAt = -Infinity;
    this.lastReason = null;
    this.healthySince = null;
  }

  observe(now: number, received: number, sent: number | null, broken: boolean): string | null {
    const previous = this.previous;
    this.previous = { received, sent };
    if (
      previous &&
      (received < previous.received ||
        (sent !== null && previous.sent !== null && sent < previous.sent))
    ) {
      this.resetSample();
      this.previous = { received, sent };
      return null;
    }
    if (sent !== null && previous?.sent != null && sent > previous.sent)
      this.remoteProgressAt = now;
    const receiving = previous !== undefined && received > previous.received;
    const reason = broken
      ? 'audio-track-or-negotiation'
      : previous && !receiving && sent !== null && now - this.remoteProgressAt < 6000
        ? 'inbound-audio-stalled'
        : null;
    if (reason === null) {
      this.suspectSince = null;
      this.lastReason = null;
      if (receiving) {
        this.healthySince ??= now;
        if (now - this.healthySince >= 60000) {
          this.failures = 0;
          this.nextRepairAt = 0;
        }
      } else this.healthySince = null;
      return null;
    }
    this.healthySince = null;
    if (reason !== this.lastReason) {
      this.suspectSince = now;
      this.lastReason = reason;
    }
    if (now - (this.suspectSince ?? now) < 8000 || now < this.nextRepairAt) return null;
    this.nextRepairAt = now + Math.min(120000, 30000 * 2 ** Math.min(this.failures, 2));
    this.failures = Math.min(this.failures + 1, 3);
    this.suspectSince = now;
    return reason;
  }
}

export const VOICE_HEALTH_CHANNEL = 'mctier-voice-health-v1';
export function parseVoiceHealth(data: unknown): number | null {
  if (typeof data !== 'string' || data.length > 128) return null;
  try {
    const value = JSON.parse(data);
    return value?.v === 1 && Number.isSafeInteger(value.packets) && value.packets >= 0
      ? value.packets
      : null;
  } catch {
    return null;
  }
}

File: MCTier-Android/app/src/main/java/top/pmh13/mctier/network/VoiceHealth.kt
package top.pmh13.mctier.network

import org.json.JSONObject

internal fun shouldQueueVoiceIce(remoteDescriptionSet: Boolean, add: () -> Boolean): Boolean =
    !remoteDescriptionSet || !runCatching(add).getOrDefault(false)

internal fun parseVoiceHealth(data: String): Long? {
    if (data.length > 128) return null
    return runCatching {
        val value = JSONObject(data)
        val packets = value.opt("packets") as? Number ?: return null
        val count = packets.toDouble()
        if (value.opt("v") != 1 || !count.isFinite() || count < 0 || count > 9007199254740991.0 || count != packets.toLong().toDouble()) null else packets.toLong()
    }.getOrNull()
}

/** Silence/DTX is healthy. A stalled receiver needs evidence of ongoing remote RTP. */
internal class VoiceHealth {
    private var previousReceived: Long? = null
    private var previousSent: Long? = null
    private var suspectSince: Long? = null
    private var remoteProgressAt: Long? = null
    private var lastReason: String? = null
    private var nextRepairAt = 0L
    private var failures = 0
    private var healthySince: Long? = null

    fun resetSample() {
        previousReceived = null; previousSent = null; suspectSince = null
        remoteProgressAt = null; lastReason = null; healthySince = null
    }

    fun observe(now: Long, received: Long, sent: Long?, broken: Boolean): String? {
        val previous = previousReceived
        val oldSent = previousSent
        previousReceived = received; previousSent = sent
        if ((previous != null && received < previous) || (sent != null && oldSent != null && sent < oldSent)) {
            resetSample(); previousReceived = received; previousSent = sent
            return null
        }
        if (sent != null && oldSent != null && sent > oldSent) remoteProgressAt = now
        val receiving = previous != null && received > previous
        val reason = when {
            broken -> "audio-track-or-negotiation"
            previous != null && !receiving && sent != null && remoteProgressAt?.let { now - it < 6000 } == true -> "inbound-audio-stalled"
            else -> null
        }
        if (reason == null) {
            suspectSince = null; lastReason = null
            if (receiving) {
                if (healthySince == null) healthySince = now
                if (now - healthySince!! >= 60000) { failures = 0; nextRepairAt = 0L }
            } else healthySince = null
            return null
        }
        healthySince = null
        if (reason != lastReason) { suspectSince = now; lastReason = reason }
        if (now - (suspectSince ?: now) < 8000 || now < nextRepairAt) return null
        nextRepairAt = now + minOf(120000L, 30000L * (1L shl failures.coerceAtMost(2)))
        failures = (failures + 1).coerceAtMost(3)
        suspectSince = now
        return reason
    }
}
